optimization/sggp/sggp-algorithm-with-test.metta

```diff
@@ -0,0 +1,597 @@
+;; it works by generating trees from the stochastic grammar values
+;; it filters the trees as good trees and bad trees
+;; it reinforces the stochastic grammar based on the two classified syntax trees
+;; The threshold is specified to the main loop which is given to the generic function 
+;; called filterlist.
+(= (sggpHelper $initSymbol $learnRate $mutationRate $thresholdProb $maxRecur $explorationCoeff $maxDepth $scoreFunc)
+    (let () (println! ($maxRecur))
+    (if (> $maxRecur 0)
+        (let* (
+            ($generatedInstances (collapse (unique (generateTrees $initSymbol &grammarSpace &stochasticGrammarSpace $explorationCoeff $maxDepth))))
+            ($_ (println! (generatedInstances ==> $generatedInstances)))
+            ;;($collapsedValues (collapse $generatedInstances))
+            ($goodTrees (collapse (filter' >= $scoreFunc (superpose $generatedInstances) $thresholdProb)))
+            ($badTrees (collapse (filter' < $scoreFunc (superpose $generatedInstances) $thresholdProb)))
+            ($_ (println! (goodTrees $goodTrees badTrees $badTrees)))
+            ($lenGood (len $goodTrees))
+            ($lenBad (len $badTrees))
+            ($reinforced (collapse (reinforceTreeRulesIterator  $goodTrees $learnRate $lenGood &grammarSpace &stochasticGrammarSpace)))
+            ($punished (collapse (punishTreeRulesIterator  $badTrees $learnRate $lenBad &grammarSpace &stochasticGrammarSpace)))
+        )
+            (let () (println! (goodTrees ==> $goodTrees))
+                (sggpHelper $initSymbol $learnRate $mutationRate $thresholdProb (- $maxRecur 1) $explorationCoeff $maxDepth $scoreFunc)
+            )
+            
+        )
+        ;; (generateInstances junctor &grammarSpace &stochasticGrammarSpace $maxDepth)
+        (collapse (unique (generateTrees $initSymbol &grammarSpace &stochasticGrammarSpace $explorationCoeff $maxDepth)))
+    )
+)
+)
+;;Helper for a reinforceGrammarBased on Tree
+(=(reinforceTreeHelper $distributedProdRules $learnRate $cardinality $stocSpace)
+    (if (== $distributedTrees ())
+        ()
+        (let* (
+            (($head $tail) (decons $distributedProdRules))
+            ($reinforced (reinforceProd $head $learnRate $cardinality $stocSpace))
+            ($stocAtoms (collapse (get-atoms $stocSpace)))
+            (() (println! (stocAtoms $stocAtoms head $head tail $tail)))
+        )
+            (reinforceTreeHelper $tail $learnRate $cardinality $stocSpace)
+        )
+    )
+)
+;; $helper for punish tree helper
+(=(punishTreeHelper $distributedProdRules $learnRate $cardinality $stocSpace)
+    (if (== $distributedTrees ())
+        ()
+        (let* (
+            (($head $tail) (decons $distributedProdRules))
+            ($_ (punishProd $head $learnRate $cardinality $stocSpace))
+        )
+            (punishTreeHelper $tail $learnRate $cardinality $stocSpace)
+        )
+    )
+)
+
+(= (reinforceTreeRulesIterator $setOfTrees $learnRate $cardinality  $grammarSpace $stocSpace)
```

Comment:
Is this update non determinism to avoid conflict when updating the weight of production rule?